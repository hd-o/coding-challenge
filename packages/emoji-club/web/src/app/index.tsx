import { themes } from '/src/style/theme'
import { FC, useContext, useEffect, useState } from 'react'
import { IntlProvider, useIntl } from 'react-intl'
import { ThemeProvider } from '@mui/material'
import { NextHeadCtx } from '../pkg/next/Head'
import { useThemeTypeSubject } from '../style/theme/type/subject'
import { intlIds } from '../util/intl-messages'
import { useResolveBehaviorState } from '../util/use-resolve-behavior-state'

export interface IntlMessages {
  [id: string]: string
}

export interface Intl {
  locale: string
  messages: IntlMessages
}

interface Props {
  Component: FC
  intl: Intl
}

const Content: FC<Props> = (props) => {
  const Head = useContext(NextHeadCtx)
  const intl = useIntl()

  // Disable Content SSR
  // Prevent React hydration mismatch because of app
  // using values from localStorage for initial render
  const [isClient, setIsClient] = useState(false)
  useEffect(() => setIsClient(true), [setIsClient])
  const Component = isClient ? props.Component : () => null

  return (
    <>
      <Head>
        <title>
          {intl.formatMessage({ id: intlIds.title })}
        </title>
        <meta
          name='description'
          content={intl.formatMessage({ id: intlIds.description })}
        />
        <link rel="icon" type="image/x-icon" href="/favicon.png" />
      </Head>
      <Component />
    </>
  )
}

export const App: FC<Props> = (props) => {
  const [themeType] = useResolveBehaviorState(useThemeTypeSubject)

  return (
    <IntlProvider messages={props.intl.messages} locale={props.intl.locale}>
      <ThemeProvider theme={themes[themeType]}>
        <Content {...props} />
      </ThemeProvider>
    </IntlProvider>
  )
}
