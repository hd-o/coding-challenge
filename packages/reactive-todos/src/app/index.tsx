import { FC } from 'react'
import { TodoInput } from '../todo/input'
import { TodoList } from '../todo/list'

export const App: FC = () => {
  return (
    <div style={{ padding: '5px 0 0 20px' }}>
      <h2>To-Dos</h2>
      <TodoInput />
      <TodoList />
    </div>
  )
}
