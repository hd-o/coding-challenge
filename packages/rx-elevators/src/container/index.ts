import { Container } from 'inversify'
import { createContext } from 'react'

export const container = new Container()
export const ContainerCtx = createContext(container)
