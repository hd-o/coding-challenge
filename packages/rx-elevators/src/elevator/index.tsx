import { createContext, FC, useContext } from 'react'
import { useStreamFn } from '../util/useStreamFn'
import { ElevatorCarCtx } from './car'
import { ElevatorDoorCtx } from './door'
import { ElevatorDoorState$ } from './door/state/stream'
import { ElevatorId } from './id'
import { ElevatorPanelCtx } from './panel'
import { ElevatorPosition$ } from './position/stream'
import { elevatorQueueItemCategories } from './queue/item/category'
import { useElevatorQueuePair$ } from './queue/pair/stream'

interface Props {
  doorState$: ElevatorDoorState$
  id: ElevatorId
  index: number
  position$: ElevatorPosition$
}

const Elevator: FC<Props> = (props) => {
  const ElevatorCar = useContext(ElevatorCarCtx)
  const ElevatorDoor = useContext(ElevatorDoorCtx)
  const ElevatorPanel = useContext(ElevatorPanelCtx)

  const queuePairs = useStreamFn(useElevatorQueuePair$)
  const isMoving = queuePairs
    .find(({ elevator }) => elevator === props.id)
    ?.queue.first()?.category === elevatorQueueItemCategories.Floor

  const idBlock = `elevator-${props.index}`
  const idModifier = isMoving ? '-moving' : ''

  return (
    <th
      data-testid={`${idBlock}${idModifier}`}
      style={{ textAlign: 'center', position: 'relative', overflow: 'visible' }}
    >
      <ElevatorPanel
        data-testid={idBlock}
        elevator={props.id}
      />
      <ElevatorCar position$={props.position$}>
        <ElevatorDoor
          data-testid={`${idBlock}-door`}
          state$={props.doorState$}
        />
      </ElevatorCar>
    </th>
  )
}

export const ElevatorCtx = createContext(Elevator)
