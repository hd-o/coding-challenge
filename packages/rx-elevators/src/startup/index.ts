import { Subscription } from 'rxjs'
import { useElevatorPositionPair$ } from '../elevator/position/pair/stream'
import { useElevatorPositionPair$Proxy } from '../elevator/position/pair/stream/proxy'
import { useElevatorQueuePair$ } from '../elevator/queue/pair/stream'
import { useElevatorQueuePair$Proxy } from '../elevator/queue/pair/stream/proxy'
import { Use } from '../util/resolve'

type Startup = () => Subscription

export const useStartup: Use<Startup> = (resolve) => {
  const elevatorQueuePair$ = resolve(useElevatorQueuePair$)
  const elevatorQueuePair$Proxy = resolve(useElevatorQueuePair$Proxy)
  const elevatorPositionPair$ = resolve(useElevatorPositionPair$)
  const elevatorPositionPair$Proxy = resolve(useElevatorPositionPair$Proxy)

  const startup = (): Subscription => {
    const modelSub = new Subscription()
    modelSub.add(elevatorQueuePair$.subscribe(elevatorQueuePair$Proxy))
    modelSub.add(elevatorPositionPair$.subscribe(elevatorPositionPair$Proxy))
    return modelSub
  }

  return startup
}
