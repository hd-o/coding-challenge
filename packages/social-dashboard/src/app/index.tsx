import { themes } from '/src/styles/theme'
import { AppProps } from 'next/app'
import Head from 'next/head'
import { FC } from 'react'
import { Cookies, CookiesProvider } from 'react-cookie'
import { DndProvider } from 'react-dnd'
import { HTML5Backend } from 'react-dnd-html5-backend'
import { IntlProvider } from 'react-intl'
import { CssBaseline, ThemeProvider } from '@mui/material'
import { useAppThemeState } from '../styles/theme/state'

export interface IntlMessages {
  [id: string]: string
}

export interface AppContainerProps {
  Component: AppProps['Component']
  cookie: string
  intl: {
    locale: string
    messages: IntlMessages
  }
  pageProps: any
}

const AppProviders: FC<AppContainerProps> = (props) => {
  const [theme] = useAppThemeState()
  return (
    <DndProvider backend={HTML5Backend}>
      <IntlProvider messages={props.intl.messages} locale={props.intl.locale}>
        <ThemeProvider theme={themes[theme]}>
          {props.children}
        </ThemeProvider>
      </IntlProvider>
    </DndProvider>
  )
}

export const AppContainer: FC<AppContainerProps> = (props) => {
  const { Component, pageProps } = props
  return (
    <CookiesProvider cookies={new Cookies(props.cookie)} >
      <AppProviders {...props}>
        <Head>
          <meta name='viewport' content='initial-scale=1, width=device-width' />
        </Head>
        <CssBaseline />
        <Component {...pageProps} />
      </AppProviders>
    </CookiesProvider>
  )
}
