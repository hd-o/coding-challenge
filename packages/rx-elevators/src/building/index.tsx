import { createContext, FC, useContext } from 'react'
import { ElevatorContainerCtx } from '../elevator/container'
import { FloorContainerCtx } from '../floor/container'
import { floorHeight } from '../floor/height'

const Building: FC = () => {
  const ElevatorContainer = useContext(ElevatorContainerCtx)
  const FloorContainer = useContext(FloorContainerCtx)

  return (
    <table className='ui compact celled definition unstackable table'>
      <tbody>
        <FloorContainer />
      </tbody>
      <tfoot className='full-width'>
        <tr style={{ height: floorHeight }}>
          <th />
          <ElevatorContainer />
        </tr>
      </tfoot>
    </table>
  )
}

export const BuildingCtx = createContext(Building)
