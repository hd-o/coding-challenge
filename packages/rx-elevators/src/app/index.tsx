import { FC, useContext, useEffect, useMemo } from 'react'
import { Subscription } from 'rxjs'
import { BuildingCtx } from '../building'
import { useStartup } from '../startup'
import { useResolve } from '../util/useResolve'

interface Props {
  TestCtrl?: FC
  subscription?: Subscription
}

export const App: FC<Props> = (props) => {
  const Building = useContext(BuildingCtx)
  const TestCtrl = props.TestCtrl ?? (() => <></>)
  const startup = useResolve(useStartup)

  /**
   * Startup subscription needs to happen before rendering
   * child components, so subscribers to shared observables
   * receive the first emitted value
   */
  const subscription = useMemo(
    () => props.subscription ?? startup(),
    [props, startup],
  )

  useEffect(() => () => subscription.unsubscribe())

  return <>
    <div style={{ paddingTop: 20, width: 800, margin: '0 auto' }}>
      <div className='ui container'>
        <Building />
      </div>
    </div>
    <TestCtrl />
  </>
}
