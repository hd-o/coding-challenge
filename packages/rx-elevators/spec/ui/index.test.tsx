/**
 * @jest-environment jsdom
 */
import '@testing-library/jest-dom'
import { App } from '/src/app'
import { render } from '@testing-library/react'

describe('UI render', () => {
  test('App snapshot', async () => {
    const app = render(<App />)
    expect(app.asFragment()).toMatchSnapshot('app-render')
  })
})
