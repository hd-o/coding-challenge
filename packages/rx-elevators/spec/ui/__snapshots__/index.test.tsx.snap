// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI render App snapshot: app-render 1`] = `
<DocumentFragment>
  <div
    style="padding-top: 20px; width: 800px; margin: 0px auto;"
  >
    <div
      class="ui container"
    >
      <table
        class="ui compact celled definition unstackable table"
      >
        <tbody>
          <tr
            style="height: 100px;"
          >
            <td
              class="two wide"
              style="text-align: center;"
            >
              <button
                class="ui button info"
                data-testid="floor-4-caller"
                style="padding: 10px; margin-top: 5px;"
              >
                4
              </button>
            </td>
            <td />
            <td />
            <td />
          </tr>
          <tr
            style="height: 100px;"
          >
            <td
              class="two wide"
              style="text-align: center;"
            >
              <button
                class="ui button info"
                data-testid="floor-3-caller"
                style="padding: 10px; margin-top: 5px;"
              >
                3
              </button>
            </td>
            <td />
            <td />
            <td />
          </tr>
          <tr
            style="height: 100px;"
          >
            <td
              class="two wide"
              style="text-align: center;"
            >
              <button
                class="ui button info"
                data-testid="floor-2-caller"
                style="padding: 10px; margin-top: 5px;"
              >
                2
              </button>
            </td>
            <td />
            <td />
            <td />
          </tr>
          <tr
            style="height: 100px;"
          >
            <td
              class="two wide"
              style="text-align: center;"
            >
              <button
                class="ui button info"
                data-testid="floor-1-caller"
                style="padding: 10px; margin-top: 5px;"
              >
                1
              </button>
            </td>
            <td />
            <td />
            <td />
          </tr>
          <tr
            style="height: 100px;"
          >
            <td
              class="two wide"
              style="text-align: center;"
            >
              <button
                class="ui button info"
                data-testid="floor-0-caller"
                style="padding: 10px; margin-top: 5px;"
              >
                0
              </button>
            </td>
            <td />
            <td />
            <td />
          </tr>
        </tbody>
        <tfoot
          class="full-width"
        >
          <tr
            style="height: 100px;"
          >
            <th />
            <th
              data-testid="elevator-0"
              style="text-align: center; position: relative; overflow: visible;"
            >
              <div>
                <button
                  class="ui button info"
                  data-testid="elevator-0-button-floor-0"
                  style="padding: 10px; margin-top: 5px;"
                >
                  0
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-0-button-floor-1"
                  style="padding: 10px; margin-top: 5px;"
                >
                  1
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-0-button-floor-2"
                  style="padding: 10px; margin-top: 5px;"
                >
                  2
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-0-button-floor-3"
                  style="padding: 10px; margin-top: 5px;"
                >
                  3
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-0-button-floor-4"
                  style="padding: 10px; margin-top: 5px;"
                >
                  4
                </button>
              </div>
              <div
                style="height: 100px; width: 100%; background: rgb(187, 187, 187); position: absolute; bottom: 100px; left: 0px;"
              >
                <div
                  data-testid="elevator-0-door-closed"
                  style="display: flex; height: 100%; width: 100%; justify-content: space-between;"
                >
                  <div
                    style="background: rgb(229, 229, 229); height: 100%; width: 50%;"
                  />
                  <div
                    style="background: rgb(229, 229, 229); height: 100%; width: 50%;"
                  />
                </div>
              </div>
            </th>
            <th
              data-testid="elevator-1"
              style="text-align: center; position: relative; overflow: visible;"
            >
              <div>
                <button
                  class="ui button info"
                  data-testid="elevator-1-button-floor-0"
                  style="padding: 10px; margin-top: 5px;"
                >
                  0
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-1-button-floor-1"
                  style="padding: 10px; margin-top: 5px;"
                >
                  1
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-1-button-floor-2"
                  style="padding: 10px; margin-top: 5px;"
                >
                  2
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-1-button-floor-3"
                  style="padding: 10px; margin-top: 5px;"
                >
                  3
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-1-button-floor-4"
                  style="padding: 10px; margin-top: 5px;"
                >
                  4
                </button>
              </div>
              <div
                style="height: 100px; width: 100%; background: rgb(187, 187, 187); position: absolute; bottom: 100px; left: 0px;"
              >
                <div
                  data-testid="elevator-1-door-closed"
                  style="display: flex; height: 100%; width: 100%; justify-content: space-between;"
                >
                  <div
                    style="background: rgb(229, 229, 229); height: 100%; width: 50%;"
                  />
                  <div
                    style="background: rgb(229, 229, 229); height: 100%; width: 50%;"
                  />
                </div>
              </div>
            </th>
            <th
              data-testid="elevator-2"
              style="text-align: center; position: relative; overflow: visible;"
            >
              <div>
                <button
                  class="ui button info"
                  data-testid="elevator-2-button-floor-0"
                  style="padding: 10px; margin-top: 5px;"
                >
                  0
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-2-button-floor-1"
                  style="padding: 10px; margin-top: 5px;"
                >
                  1
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-2-button-floor-2"
                  style="padding: 10px; margin-top: 5px;"
                >
                  2
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-2-button-floor-3"
                  style="padding: 10px; margin-top: 5px;"
                >
                  3
                </button>
                <button
                  class="ui button info"
                  data-testid="elevator-2-button-floor-4"
                  style="padding: 10px; margin-top: 5px;"
                >
                  4
                </button>
              </div>
              <div
                style="height: 100px; width: 100%; background: rgb(187, 187, 187); position: absolute; bottom: 100px; left: 0px;"
              >
                <div
                  data-testid="elevator-2-door-closed"
                  style="display: flex; height: 100%; width: 100%; justify-content: space-between;"
                >
                  <div
                    style="background: rgb(229, 229, 229); height: 100%; width: 50%;"
                  />
                  <div
                    style="background: rgb(229, 229, 229); height: 100%; width: 50%;"
                  />
                </div>
              </div>
            </th>
          </tr>
        </tfoot>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
